// web3x/contract/contract.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const formatters_1 = require("../formatters");
const subscriptions_1 = require("../subscriptions");
const utils_1 = require("../utils");
const tx_1 = require("./tx");
const tx_deploy_1 = require("./tx-deploy");
/**
 * Should be called to create new contract instance
 *
 * @method Contract
 * @constructor
 * @param {Array} jsonInterface
 * @param {String} address
 * @param {Object} options
 */
class Contract {
    constructor(eth, contractAbi, address, defaultOptions = {}) {
        this.eth = eth;
        this.contractAbi = contractAbi;
        this.address = address;
        this.defaultOptions = defaultOptions;
        this.linkTable = {};
        this.methods = this.buildMethods();
        this.events = this.buildEvents();
    }
    link(name, address) {
        this.linkTable[name] = address;
    }
    deployBytecode(data, ...args) {
        const linkedData = Object.entries(this.linkTable).reduce((data, [name, address]) => data.replace(new RegExp(`_+${name}_+`, 'gi'), address
            .toString()
            .slice(2)
            .toLowerCase()), data);
        if (linkedData.includes('_')) {
            throw new Error('Bytecode has not been fully linked.');
        }
        return new tx_deploy_1.TxDeploy(this.eth, this.contractAbi.ctor, this.contractAbi, utils_1.hexToBuffer(linkedData), args, this.defaultOptions, addr => (this.address = addr));
    }
    once(event, options, callback) {
        this.on(event, options, (err, res, sub) => {
            sub.unsubscribe();
            callback(err, res, sub);
        });
    }
    async getPastEvents(event, options = {}) {
        const logOptions = this.getLogOptions(event, options);
        const result = await this.eth.getPastLogs(logOptions);
        return result.map(log => this.contractAbi.decodeEvent(log));
    }
    on(event, options = {}, callback) {
        const logOptions = this.getLogOptions(event, options);
        const { fromBlock, ...subLogOptions } = logOptions;
        const params = [formatters_1.toRawLogRequest(subLogOptions)];
        const subscription = new subscriptions_1.Subscription('eth', 'logs', params, this.eth.provider, (result, sub) => {
            const output = formatters_1.fromRawLogResponse(result);
            const eventLog = this.contractAbi.decodeEvent(output);
            sub.emit(output.removed ? 'changed' : 'data', eventLog);
            if (callback) {
                callback(undefined, eventLog, sub);
            }
        }, false);
        subscription.on('error', err => {
            if (callback) {
                callback(err, undefined, subscription);
            }
        });
        if (fromBlock !== undefined) {
            this.eth
                .getPastLogs(logOptions)
                .then(logs => {
                logs.forEach(result => {
                    const output = this.contractAbi.decodeEvent(result);
                    subscription.emit('data', output);
                });
                subscription.subscribe();
            })
                .catch(err => {
                subscription.emit('error', err);
            });
        }
        else {
            subscription.subscribe();
        }
        return subscription;
    }
    executorFactory(functions) {
        return (...args) => {
            if (!this.address) {
                throw new Error('No contract address.');
            }
            const firstMatchingOverload = functions.find(f => args.length === f.numArgs());
            if (!firstMatchingOverload) {
                throw new Error(`No matching method with ${args.length} arguments for ${functions[0].name}.`);
            }
            return new tx_1.Tx(this.eth, firstMatchingOverload, this.contractAbi, this.address, args, this.defaultOptions);
        };
    }
    buildMethods() {
        const methods = {};
        this.contractAbi.functions.forEach(f => {
            const executor = this.executorFactory([f]);
            methods[f.asString()] = executor;
            methods[f.signature] = executor;
        });
        const grouped = this.contractAbi.functions.reduce((acc, method) => {
            const funcs = [...(acc[method.name] || []), method];
            return { ...acc, [method.name]: funcs };
        }, {});
        Object.entries(grouped).map(([name, funcs]) => {
            methods[name] = this.executorFactory(funcs);
        });
        return methods;
    }
    buildEvents() {
        const events = {};
        this.contractAbi.events.forEach(e => {
            const event = this.on.bind(this, e.signature);
            if (!events[e.name]) {
                events[e.name] = event;
            }
            events[e.asString()] = event;
            events[e.signature] = event;
        });
        events.allEvents = this.on.bind(this, 'allevents');
        return events;
    }
    getLogOptions(eventName = 'allevents', options) {
        if (!this.address) {
            throw new Error('No contract address.');
        }
        if (eventName.toLowerCase() === 'allevents') {
            return {
                ...options,
                address: this.address,
            };
        }
        const event = this.contractAbi.events.find(e => e.name === eventName || e.signature === '0x' + eventName.replace('0x', ''));
        if (!event) {
            throw new Error(`Event ${eventName} not found.`);
        }
        return {
            ...options,
            address: this.address,
            topics: event.getEventTopics(options.filter),
        };
    }
}
exports.Contract = Contract;
