// web3x/contract/abi/contract-abi.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require(".");
const utils_1 = require("../../utils");
class ContractAbi {
    constructor(definition) {
        this.functions = definition.filter(e => e.type === 'function').map(entry => new _1.ContractFunctionEntry(entry));
        this.events = definition.filter(e => e.type === 'event').map(entry => new _1.ContractEventEntry(entry));
        const ctor = definition.find(e => e.type === 'constructor');
        this.ctor = new _1.ContractFunctionEntry(ctor || { type: 'constructor' });
        const fallback = definition.find(e => e.type === 'fallback');
        if (fallback) {
            this.fallback = new _1.ContractFunctionEntry(fallback);
        }
    }
    findEntryForLog(log) {
        return this.events.find(abiDef => abiDef.signature === log.topics[0]);
    }
    decodeEvent(log) {
        const event = this.findEntryForLog(log);
        if (!event) {
            throw new Error(`Unable to find matching event signature for log: ${log.id}`);
        }
        return event.decodeEvent(log);
    }
    decodeFunctionData(data) {
        const funcSig = utils_1.bufferToHex(data.slice(0, 4));
        const func = this.functions.find(f => f.signature === funcSig);
        return func ? func.decodeParameters(data.slice(4)) : undefined;
    }
}
exports.ContractAbi = ContractAbi;
