// web3x/contract/abi/contract-event-entry.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("util");
const abi_coder_1 = require("../abi-coder");
const contract_entry_1 = require("./contract-entry");
class ContractEventEntry extends contract_entry_1.ContractEntry {
    constructor(entry) {
        super(entry);
        this.signature = abi_coder_1.abiCoder.encodeEventSignature(abi_coder_1.abiCoder.abiMethodToString(entry));
    }
    getEventTopics(filter = {}) {
        const topics = [];
        if (!this.entry.anonymous && this.signature) {
            topics.push(this.signature);
        }
        const indexedTopics = (this.entry.inputs || [])
            .filter(input => input.indexed === true)
            .map(input => {
            const value = filter[input.name];
            if (!value) {
                return null;
            }
            // TODO: https://github.com/ethereum/web3.js/issues/344
            // TODO: deal properly with components
            if (util_1.isArray(value)) {
                return value.map(v => abi_coder_1.abiCoder.encodeParameter(input.type, v));
            }
            else {
                return abi_coder_1.abiCoder.encodeParameter(input.type, value);
            }
        });
        return [...topics, ...indexedTopics];
    }
    decodeEvent(log) {
        const { data = '', topics = [], ...formattedLog } = log;
        const { anonymous, inputs = [], name = '' } = this.entry;
        const argTopics = anonymous ? topics : topics.slice(1);
        const returnValues = abi_coder_1.abiCoder.decodeLog(inputs, data, argTopics);
        delete returnValues.__length__;
        return {
            ...formattedLog,
            event: name,
            returnValues,
            signature: anonymous || !topics[0] ? null : topics[0],
            raw: {
                data,
                topics,
            },
        };
    }
}
exports.ContractEventEntry = ContractEventEntry;
