// web3x/providers/http.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const http_1 = tslib_1.__importDefault(require("http"));
const https_1 = tslib_1.__importDefault(require("https"));
const isomorphic_fetch_1 = tslib_1.__importDefault(require("isomorphic-fetch"));
const legacy_provider_adapter_1 = require("./legacy-provider-adapter");
class HttpProvider extends legacy_provider_adapter_1.LegacyProviderAdapter {
    constructor(host, options = {}) {
        super(new LegacyHttpProvider(host, options));
    }
}
exports.HttpProvider = HttpProvider;
class LegacyHttpProvider {
    constructor(host, options = {}) {
        this.host = host;
        this.options = options;
        this.host = host || 'http://localhost:8545';
        if (options.keepAlive) {
            this.options.agent = /^https/.test(this.host)
                ? new https_1.default.Agent({ keepAlive: true })
                : new http_1.default.Agent({ keepAlive: true });
        }
    }
    send(payload, callback) {
        isomorphic_fetch_1.default(this.host, {
            ...this.options,
            method: 'POST',
            credentials: 'include',
            headers: {
                ...this.options.headers,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(payload),
        })
            .then(response => response.json())
            .then(json => callback(undefined, json))
            .catch(callback);
    }
    disconnect() { }
}
