// web3x/eth/eth.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const providers_1 = require("../providers");
const eth_request_payloads_1 = require("./eth-request-payloads");
const send_tx_1 = require("./send-tx");
const logs_1 = require("./subscriptions/logs");
const new_heads_1 = require("./subscriptions/new-heads");
const new_pending_transactions_1 = require("./subscriptions/new-pending-transactions");
const syncing_1 = require("./subscriptions/syncing");
class Eth {
    constructor(provider) {
        this.provider = provider;
        this.request = new eth_request_payloads_1.EthRequestPayloads(undefined, 'latest');
    }
    static fromCurrentProvider() {
        if (typeof web3 === 'undefined') {
            return;
        }
        const provider = web3.currentProvider || web3.ethereumProvider;
        if (!provider) {
            return;
        }
        return new Eth(new providers_1.LegacyProviderAdapter(provider));
    }
    get defaultFromAddress() {
        return this.request.defaultFromAddress;
    }
    set defaultFromAddress(address) {
        this.request.defaultFromAddress = address;
    }
    async send({ method, params, format }) {
        return format(await this.provider.send(method, params));
    }
    async getId() {
        return await this.send(this.request.getId());
    }
    async getNodeInfo() {
        return await this.send(this.request.getNodeInfo());
    }
    async getProtocolVersion() {
        return await this.send(this.request.getProtocolVersion());
    }
    async getCoinbase() {
        return await this.send(this.request.getCoinbase());
    }
    async isMining() {
        return await this.send(this.request.isMining());
    }
    async getHashrate() {
        return await this.send(this.request.getHashrate());
    }
    async isSyncing() {
        return await this.send(this.request.isSyncing());
    }
    async getGasPrice() {
        return await this.send(this.request.getGasPrice());
    }
    async getAccounts() {
        return await this.send(this.request.getAccounts());
    }
    async getBlockNumber() {
        return await this.send(this.request.getBlockNumber());
    }
    async getBalance(address, block) {
        return await this.send(this.request.getBalance(address, block));
    }
    async getStorageAt(address, position, block) {
        return await this.send(this.request.getStorageAt(address, position, block));
    }
    async getCode(address, block) {
        return await this.send(this.request.getCode(address, block));
    }
    async getBlock(block, returnTxs) {
        return await this.send(this.request.getBlock(block, returnTxs));
    }
    async getUncle(block, uncleIndex, returnTxs) {
        return await this.send(this.request.getUncle(block, uncleIndex, returnTxs));
    }
    async getBlockTransactionCount(block) {
        return await this.send(this.request.getBlockTransactionCount(block));
    }
    async getBlockUncleCount(block) {
        return await this.send(this.request.getBlockUncleCount(block));
    }
    async getTransaction(hash) {
        return await this.send(this.request.getTransaction(hash));
    }
    async getTransactionFromBlock(block, index) {
        return await this.send(this.request.getTransactionFromBlock(block, index));
    }
    async getTransactionReceipt(txHash) {
        return await this.send(this.request.getTransactionReceipt(txHash));
    }
    async getTransactionCount(address, block) {
        return await this.send(this.request.getTransactionCount(address, block));
    }
    async signTransaction(tx) {
        return await this.send(this.request.signTransaction(tx));
    }
    sendSignedTransaction(data) {
        const { method, params } = this.request.sendSignedTransaction(data);
        const txHashPromise = this.provider.send(method, params);
        return new send_tx_1.SentTransaction(this, txHashPromise);
    }
    sendTransaction(tx) {
        const promise = new Promise(async (resolve, reject) => {
            try {
                if (!tx.gasPrice) {
                    tx.gasPrice = await this.getGasPrice();
                }
                const account = this.getAccount(tx.from);
                if (!account) {
                    const { method, params, format } = this.request.sendTransaction(tx);
                    const txHash = format(await this.provider.send(method, params));
                    resolve(txHash);
                }
                else {
                    const { from, ...fromlessTx } = tx;
                    const signedTx = await account.signTransaction(fromlessTx, this);
                    const { method, params, format } = this.request.sendSignedTransaction(signedTx.rawTransaction);
                    const txHash = format(await this.provider.send(method, params));
                    resolve(txHash);
                }
            }
            catch (err) {
                reject(err);
            }
        });
        return new send_tx_1.SentTransaction(this, promise);
    }
    getAccount(address) {
        address = address || this.defaultFromAddress;
        if (this.wallet && address) {
            return this.wallet.get(address);
        }
    }
    async sign(address, dataToSign) {
        const account = this.getAccount(address);
        if (!account) {
            return await this.send(this.request.sign(address, dataToSign));
        }
        else {
            const sig = account.sign(dataToSign);
            return sig.signature;
        }
    }
    async signTypedData(address, dataToSign) {
        return await this.send(this.request.signTypedData(address, dataToSign));
    }
    async call(tx, block) {
        return await this.send(this.request.call(tx, block));
    }
    async estimateGas(tx) {
        return await this.send(this.request.estimateGas(tx));
    }
    async submitWork(nonce, powHash, digest) {
        return await this.send(this.request.submitWork(nonce, powHash, digest));
    }
    async getWork() {
        return await this.send(this.request.getWork());
    }
    async getPastLogs(options) {
        return await this.send(this.request.getPastLogs(options));
    }
    subscribe(type, ...args) {
        switch (type) {
            case 'logs':
                return logs_1.subscribeForLogs(this, ...args);
            case 'syncing':
                return syncing_1.subscribeForSyncing(this.provider);
            case 'newBlockHeaders':
                return new_heads_1.subscribeForNewHeads(this.provider);
            case 'pendingTransactions':
                return new_pending_transactions_1.subscribeForNewPendingTransactions(this.provider);
            default:
                throw new Error(`Unknown subscription type: ${type}`);
        }
    }
}
exports.Eth = Eth;
