// web3x/formatters/estimate-request-formatter.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const address_1 = require("../address");
const utils_1 = require("../utils");
function toRawEstimateRequest(tx) {
    const { from, to, gas, gasPrice, value, data } = tx;
    return {
        from: from ? from.toString().toLowerCase() : undefined,
        to: to ? to.toString().toLowerCase() : undefined,
        data: data ? utils_1.bufferToHex(data) : undefined,
        value: value ? utils_1.numberToHex(value) : undefined,
        gas: gas ? utils_1.numberToHex(gas) : undefined,
        gasPrice: gasPrice ? utils_1.numberToHex(gasPrice) : undefined,
    };
}
exports.toRawEstimateRequest = toRawEstimateRequest;
function fromRawEstimateRequest(tx) {
    const { from, to, gas, gasPrice, value, data } = tx;
    return {
        from: from ? address_1.Address.fromString(from) : undefined,
        to: to ? address_1.Address.fromString(to) : undefined,
        data: data ? utils_1.hexToBuffer(data) : undefined,
        value: value ? utils_1.hexToNumberString(value) : undefined,
        gas: gas ? utils_1.hexToNumberString(gas) : undefined,
        gasPrice: gasPrice ? utils_1.hexToNumberString(gasPrice) : undefined,
    };
}
exports.fromRawEstimateRequest = fromRawEstimateRequest;
