// web3x/personal/personal.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const personal_request_payloads_1 = require("./personal-request-payloads");
class Personal {
    constructor(provider) {
        this.provider = provider;
        this.request = new personal_request_payloads_1.PersonalRequestPayloads();
    }
    async send({ method, params, format }) {
        return format(await this.provider.send(method, params));
    }
    async getAccounts() {
        const payload = this.request.getAccounts();
        return await this.send(payload);
    }
    async newAccount(password) {
        const payload = this.request.newAccount(password);
        return await this.send(payload);
    }
    async unlockAccount(address, password, duration) {
        const payload = this.request.unlockAccount(address, password, duration);
        return await this.send(payload);
    }
    async lockAccount(address) {
        const payload = this.request.lockAccount(address);
        return await this.send(payload);
    }
    async importRawKey(privateKey, password) {
        const payload = this.request.importRawKey(privateKey, password);
        return await this.send(payload);
    }
    async sendTransaction(tx, password) {
        const payload = this.request.sendTransaction(tx, password);
        return await this.send(payload);
    }
    async signTransaction(tx, password) {
        const payload = this.request.signTransaction(tx, password);
        return await this.send(payload);
    }
    async sign(message, address, password) {
        const payload = this.request.sign(message, address, password);
        return await this.send(payload);
    }
    async ecRecover(message, signedData) {
        const payload = this.request.ecRecover(message, signedData);
        return await this.send(payload);
    }
}
exports.Personal = Personal;
